### ISSUE TITLE
about toy experiment

### ISSUE THREAD

**Issue by modyu-liu** [LTH14/JiT#9]
Thank you very much for your insightful work. I encountered a difficulty while reproducing your toy experiment.

Using the settings provided in your paper, the 16-channel and 512-channel x_pred and v_pred. I obtained look like the figures below.

<img width="1000" height="1000" alt="Image" src="https://github.com/user-attachments/assets/1955c50e-f07e-44dd-b231-5a777c4bc877" />
<img width="1000" height="1000" alt="Image" src="https://github.com/user-attachments/assets/7dfc9ce2-8eb8-4938-a3c1-052b502bdb9d" />
<img width="1000" height="1000" alt="Image" src="https://github.com/user-attachments/assets/51cbdbf2-4398-4028-aa7f-91a648bf44e0" />
<img width="1000" height="1000" alt="Image" src="https://github.com/user-attachments/assets/d696f75d-2ec1-4625-88a7-5383fb219bf4" />

These results are quite different from Fig. 2 in your paper.

Interestingly, if I do not normalize the input data, the results become much closer to Fig. 2, as shown below.

<img width="1000" height="1000" alt="Image" src="https://github.com/user-attachments/assets/519a1abd-c722-4a7b-930a-692b5f17ba3a" />
<img width="1000" height="1000" alt="Image" src="https://github.com/user-attachments/assets/ca12c1a4-9598-40ea-bbff-4b9d56c9e4aa" />
<img width="1000" height="1000" alt="Image" src="https://github.com/user-attachments/assets/bdec7425-65a0-404d-9e8c-c89bd3dd74af" />
<img width="1000" height="1000" alt="Image" src="https://github.com/user-attachments/assets/ce78de0f-4c9a-489d-b6eb-629e2b093383" />

The training code I used is as follows:

[train_toy_exp.py](https://github.com/user-attachments/files/23628605/train_toy_exp.py)

Could you please provide some guidance on this issue?

**Comment by LTH14** [comment]
Thanks for sharing your results -- super interesting! In our toy experiment, we don’t normalize the input, and we see similar behavior to what you report without normalization. Intuitively, since v is a combination of x and eps, the scaling of x will change how much "x" the network predicts in v-prediction. When the scale of x grows and starts dominate v = x - eps, v-prediction should gradually approach x-prediction. It would be cool to see the results if you manually tune the normalization factor, and perhaps with larger data scale, the behavior of v-pred will be even closer to x-pred.

**Comment by modyu-liu** [comment]
> Thanks for sharing your results -- super interesting! In our toy experiment, we don’t normalize the input, and we see similar behavior to what you report without normalization. Intuitively, since v is a combination of x and eps, the scaling of x will change how much "x" the network predicts in v-prediction. When the scale of x grows and starts dominate v = x - eps, v-prediction should gradually approach x-prediction. It would be cool to see the results if you manually tune the normalization factor, and perhaps with larger data scale, the behavior of v-pred will be even closer to x-pred.

Thank you for your reply. Indeed, regardless of whether the data is normalized or not, x-pred seems to converge better than v-pred. However, there is a strange phenomenon: after applying normalization, the outputs of x-pred exhibit a noticeable shift, whereas without normalization it fits the data very well. This seems to suggest that x-pred is highly sensitive to whether the data is normalized.

What also confuses me is that previous community discussions often suggest that the data should ideally follow an N(0, I) distribution. But based on toy experiments, it appears that the further the data deviates from N(0, I), the better x-pred performs.

**Comment by mmjwxbc** [comment]
```python
import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import TensorDataset, DataLoader
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt
import numpy as np
from tqdm import tqdm

class MLP5(nn.Module):
    def __init__(self, in_dim, hidden_dim, out_dim):
        super().__init__()
        layers = []
        dims = [in_dim] + [hidden_dim] * 5 + [out_dim]
        for i in range(len(dims) - 2):
            layers.append(nn.Linear(dims[i], dims[i + 1]))
            layers.append(nn.ReLU(inplace=True))
        layers.append(nn.Linear(dims[-2], dims[-1]))
        self.net = nn.Sequential(*layers)

    def forward(self, x, t):
        """
        x: [B, D]
        t: [B] or [B,1]
        """
        if t.dim() == 1:
            t = t.unsqueeze(-1)
        inp = torch.cat([x, t], dim=-1)
        return self.net(inp)


def get_swiss_roll(n_samples=2000, noise=0.0):
    t = 1.5 * np.pi * (1 + 2 * np.random.rand(n_samples, 1))
    x = t * np.cos(t)
    y = t * np.sin(t)
    X = np.concatenate([x, y], axis=1)
    X += noise * np.random.randn(n_samples, 2)
    X = (X - X.mean(axis=0)) / X.std(axis=0) 
    return torch.from_numpy(X).float()

@torch.no_grad()
def sample_ode(net, dim, device, pred_type, steps=100):
    """
    z_next = z + (t_next - t) * v_pred
    """
    net.eval()
    N_sample = 2000
    z = torch.randn(N_sample, dim).to(device)
    
    times = torch.linspace(0, 1, steps + 1).to(device)
    
    for i in range(steps):
        t_current = times[i]
        t_next = times[i+1]
        dt = t_next - t_current
        
        t_input = torch.ones(N_sample, 1).to(device) * t_current
        
        pred = net(z, t_input)

        denom = (1 - t_current).clamp(min=1e-5)
        if pred_type == 'x_pred':
            v_pred = (pred - z) / denom
        elif pred_type == 'e_pred':
            v_pred = (z - pred) / t_current.clamp(min=1e-5)
        elif pred_type == 'v_pred':
            v_pred = pred

        
        z = z + dt * v_pred
        
    return z

pred_type = ['v_pred', 'x_pred', 'e_pred']
Ds = [2, 8, 16, 512] # dim 
d_intrinsic = 2      # x_hat_dim
N = 1024
num_epochs = 1000
batch_size = 64
device = 'cuda:7' if torch.cuda.is_available() else 'cpu'

trained_nets = []
projections = [] 

# 创建画布
rows = len(Ds)
cols = len(pred_type)

fig, axes = plt.subplots(
    rows, 
    cols, 
    figsize=(9, 12),
    dpi=100,
    constrained_layout=True,
    sharex=True,
    sharey=True
)
if len(Ds) == 1: axes = [axes]

print(f"Using device: {device}")

for j, type in enumerate(pred_type):
    for i, D in enumerate(Ds):
        print(f"\nTraining for D={D}...")
        
        x_flat = get_swiss_roll(N, noise=0.1) # [N, 2]
        
        if D == 2:
            P = torch.eye(2)
        else:
            P_rnd = torch.randn(D, d_intrinsic)
            Q_mat, _ = torch.linalg.qr(P_rnd) 
            P = Q_mat # [D, 2] (Assuming reduced QR)

        # x_flat: [N, 2], P: [D, 2] -> x: [N, D]
        x = x_flat @ P.T
        
        dataset = TensorDataset(x)
        dataloader = DataLoader(dataset, batch_size=batch_size, shuffle=True)

        net = MLP(D).to(device)
        opt = optim.AdamW(net.parameters(), lr=1e-3, weight_decay=1e-6)
        
        t_eps = 5e-3
        pbar =tqdm(range(num_epochs))
        for epoch in pbar:
            total_loss = 0
            for (x_batch,) in dataloader:
                x_batch = x_batch.to(device)
                batch_n = x_batch.shape[0]
                
                t = torch.rand(batch_n, device=device)
                eps = torch.randn_like(x_batch)
                
                t_b = t.view(-1, 1)
                z_t = t_b * x_batch + (1 - t_b) * eps
                
                v = (x_batch - z_t) / (1 - t_b).clamp_min(t_eps)
                pred = net(z_t, t)
                
                if type == 'x_pred':
                    v_pred = (pred - z_t) / (1 - t_b).clamp_min(t_eps)
                    # loss = ((pred  - x_batch) ** 2).mean()
                    # x_pred = pred
                elif type == 'e_pred':
                    # loss = ((pred - eps) ** 2).mean()
                    v_pred = (z_t - pred) / t_b.clamp_min(t_eps)
                elif type == 'v_pred':
                    v_pred = pred
                    # loss = ((pred - v) ** 2).mean()
                    
                    # v_pred = pred
                loss = ((v_pred - v) ** 2).mean()
                # loss = ((x_pred - x_batch) ** 2).mean()
                opt.zero_grad()
                loss.backward()
                opt.step()
                total_loss += loss.item()
                
           
            pbar.set_description(f"Epoch {epoch+1}/{num_epochs} | Loss: {total_loss / len(dataloader):.4f}")

        z_generated = sample_ode(net, D, device, type, steps=50) # [N_sample, D]
        
        z_generated_cpu = z_generated.cpu()
        P_cpu = P.cpu()
        
        z_projected = z_generated_cpu @ P_cpu
        
        # plot
        ax = axes[i][j]
        ax.scatter(z_projected[:, 0], z_projected[:, 1], s=1, alpha=0.5, color='orange')
        ax.set_title(f"$D={D}$", fontsize=14, loc='left')
        ax.set_xlim([-2.5, 2.5])
        ax.set_ylim([-2.5, 2.5])
        ax.grid(True, alpha=0.3)
        ax.set_aspect('equal')

plt.tight_layout()
plt.show()
```
Why does unifying the loss as predicting v in my method generate such strange images? The first one is v_target, and the second one is x_target.

<img width="870" height="1189" alt="Image" src="https://github.com/user-attachments/assets/6fa0cc7f-2a8f-4655-b6ca-d430546fcfa9" />

**Comment by LTH14** [comment]
@modyu-liu normalizing x is a "standard practice", but people need to perform other practices because x is structured, while eps is not -- for example, in "[this paper](https://arxiv.org/abs/2301.10972)", they need to scale x (or similarly in our paper, scale eps) to rebalance x and eps. From my understanding, there is a "sweet spot" in the relative std ratio of x and eps. This sweet spot is not always 1, and could change a lot according to the actual task.

**Comment by CyberDragon93** [comment]
![image](https://github.com/user-attachments/assets/7e4180fc-e938-4782-9a9f-e73e58901a25)

![image](https://github.com/user-attachments/assets/520614f7-bf0d-49a3-981f-de360b4eee56)

This is the result when I normalize data by dividing std and multiply by 3

**Comment by mw66** [comment]
> Thanks for sharing your results -- super interesting! In our toy experiment, we don’t normalize the input, and we see similar behavior to what you report without normalization.

This discussion is interesting,  on the other hand,  @LTH14 you may want to add your toy experiment script here in this repo, so people reading your paper can run it to compare the results published in the paper.

**Comment by zuti666** [comment]
> import torch
> import torch.nn as nn
> import torch.optim as optim
> from torch.utils.data import TensorDataset, DataLoader
> from sklearn.model_selection import train_test_split
> import matplotlib.pyplot as plt
> import numpy as np
> from tqdm import tqdm
> 
> class MLP5(nn.Module):
>     def __init__(self, in_dim, hidden_dim, out_dim):
>         super().__init__()
>         layers = []
>         dims = [in_dim] + [hidden_dim] * 5 + [out_dim]
>         for i in range(len(dims) - 2):
>             layers.append(nn.Linear(dims[i], dims[i + 1]))
>             layers.append(nn.ReLU(inplace=True))
>         layers.append(nn.Linear(dims[-2], dims[-1]))
>         self.net = nn.Sequential(*layers)
> 
>     def forward(self, x, t):
>         """
>         x: [B, D]
>         t: [B] or [B,1]
>         """
>         if t.dim() == 1:
>             t = t.unsqueeze(-1)
>         inp = torch.cat([x, t], dim=-1)
>         return self.net(inp)
> 
> 
> def get_swiss_roll(n_samples=2000, noise=0.0):
>     t = 1.5 * np.pi * (1 + 2 * np.random.rand(n_samples, 1))
>     x = t * np.cos(t)
>     y = t * np.sin(t)
>     X = np.concatenate([x, y], axis=1)
>     X += noise * np.random.randn(n_samples, 2)
>     X = (X - X.mean(axis=0)) / X.std(axis=0) 
>     return torch.from_numpy(X).float()
> 
> @torch.no_grad()
> def sample_ode(net, dim, device, pred_type, steps=100):
>     """
>     z_next = z + (t_next - t) * v_pred
>     """
>     net.eval()
>     N_sample = 2000
>     z = torch.randn(N_sample, dim).to(device)
>     
>     times = torch.linspace(0, 1, steps + 1).to(device)
>     
>     for i in range(steps):
>         t_current = times[i]
>         t_next = times[i+1]
>         dt = t_next - t_current
>         
>         t_input = torch.ones(N_sample, 1).to(device) * t_current
>         
>         pred = net(z, t_input)
> 
>         denom = (1 - t_current).clamp(min=1e-5)
>         if pred_type == 'x_pred':
>             v_pred = (pred - z) / denom
>         elif pred_type == 'e_pred':
>             v_pred = (z - pred) / t_current.clamp(min=1e-5)
>         elif pred_type == 'v_pred':
>             v_pred = pred
> 
>         
>         z = z + dt * v_pred
>         
>     return z
> 
> pred_type = ['v_pred', 'x_pred', 'e_pred']
> Ds = [2, 8, 16, 512] # dim 
> d_intrinsic = 2      # x_hat_dim
> N = 1024
> num_epochs = 1000
> batch_size = 64
> device = 'cuda:7' if torch.cuda.is_available() else 'cpu'
> 
> trained_nets = []
> projections = [] 
> 
> # 创建画布
> rows = len(Ds)
> cols = len(pred_type)
> 
> fig, axes = plt.subplots(
>     rows, 
>     cols, 
>     figsize=(9, 12),
>     dpi=100,
>     constrained_layout=True,
>     sharex=True,
>     sharey=True
> )
> if len(Ds) == 1: axes = [axes]
> 
> print(f"Using device: {device}")
> 
> for j, type in enumerate(pred_type):
>     for i, D in enumerate(Ds):
>         print(f"\nTraining for D={D}...")
>         
>         x_flat = get_swiss_roll(N, noise=0.1) # [N, 2]
>         
>         if D == 2:
>             P = torch.eye(2)
>         else:
>             P_rnd = torch.randn(D, d_intrinsic)
>             Q_mat, _ = torch.linalg.qr(P_rnd) 
>             P = Q_mat # [D, 2] (Assuming reduced QR)
> 
>         # x_flat: [N, 2], P: [D, 2] -> x: [N, D]
>         x = x_flat @ P.T
>         
>         dataset = TensorDataset(x)
>         dataloader = DataLoader(dataset, batch_size=batch_size, shuffle=True)
> 
>         net = MLP(D).to(device)
>         opt = optim.AdamW(net.parameters(), lr=1e-3, weight_decay=1e-6)
>         
>         t_eps = 5e-3
>         pbar =tqdm(range(num_epochs))
>         for epoch in pbar:
>             total_loss = 0
>             for (x_batch,) in dataloader:
>                 x_batch = x_batch.to(device)
>                 batch_n = x_batch.shape[0]
>                 
>                 t = torch.rand(batch_n, device=device)
>                 eps = torch.randn_like(x_batch)
>                 
>                 t_b = t.view(-1, 1)
>                 z_t = t_b * x_batch + (1 - t_b) * eps
>                 
>                 v = (x_batch - z_t) / (1 - t_b).clamp_min(t_eps)
>                 pred = net(z_t, t)
>                 
>                 if type == 'x_pred':
>                     v_pred = (pred - z_t) / (1 - t_b).clamp_min(t_eps)
>                     # loss = ((pred  - x_batch) ** 2).mean()
>                     # x_pred = pred
>                 elif type == 'e_pred':
>                     # loss = ((pred - eps) ** 2).mean()
>                     v_pred = (z_t - pred) / t_b.clamp_min(t_eps)
>                 elif type == 'v_pred':
>                     v_pred = pred
>                     # loss = ((pred - v) ** 2).mean()
>                     
>                     # v_pred = pred
>                 loss = ((v_pred - v) ** 2).mean()
>                 # loss = ((x_pred - x_batch) ** 2).mean()
>                 opt.zero_grad()
>                 loss.backward()
>                 opt.step()
>                 total_loss += loss.item()
>                 
>            
>             pbar.set_description(f"Epoch {epoch+1}/{num_epochs} | Loss: {total_loss / len(dataloader):.4f}")
> 
>         z_generated = sample_ode(net, D, device, type, steps=50) # [N_sample, D]
>         
>         z_generated_cpu = z_generated.cpu()
>         P_cpu = P.cpu()
>         
>         z_projected = z_generated_cpu @ P_cpu
>         
>         # plot
>         ax = axes[i][j]
>         ax.scatter(z_projected[:, 0], z_projected[:, 1], s=1, alpha=0.5, color='orange')
>         ax.set_title(f"$D={D}$", fontsize=14, loc='left')
>         ax.set_xlim([-2.5, 2.5])
>         ax.set_ylim([-2.5, 2.5])
>         ax.grid(True, alpha=0.3)
>         ax.set_aspect('equal')
> 
> plt.tight_layout()
> plt.show()
> Why does unifying the loss as predicting v in my method generate such strange images? The first one is v_target, and the second one is x_target.
> 
> <img alt="Image" width="870" height="1189" src="https://private-user-images.githubusercontent.com/93423752/516746610-6fa0cc7f-2a8f-4655-b6ca-d430546fcfa9.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjM5MTkxNjMsIm5iZiI6MTc2MzkxODg2MywicGF0aCI6Ii85MzQyMzc1Mi81MTY3NDY2MTAtNmZhMGNjN2YtMmE4Zi00NjU1LWI2Y2EtZDQzMDU0NmZjZmE5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTExMjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMTIzVDE3Mjc0M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNiMmQ1ODMzMGUyOGQxNGU0MTMyYTgxODY5YTU3NWMyNGYzOGU2NGY3MTM3ZmQyZGE4Nzc5M2M3Yzc3OWY3MzImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.k2IR3PyPTJks5Ea7SCk4JzklRrv0IIs83Kbx2ZDTtQc">

Thanks for sharing your code. But the code can not run, it  leave out the initialization of Model MLP5 .

**Comment by LTH14** [comment]
@zuti666 you can check [this reimplementation](https://github.com/SUZ-tsinghua/jit_toy_example)
